src/main/java/com/cloudogu/scm/mirror/MirrorConfigurationStore.java

```diff
@@ -63,43 +65,126 @@
+    return existingConfiguration;
   }
 
-  private void updateWithExisting(MirrorConfiguration existingConfig, MirrorConfiguration configuration) {
-    if (Strings.isNullOrEmpty(configuration.getUrl())) {
-      configuration.setUrl(existingConfig.getUrl());
+  private void updateUrl(MirrorConfiguration existingConfiguration, MirrorAccessConfiguration newMirrorAccessConfiguration) {
+    if (Strings.isNullOrEmpty(existingConfiguration.getUrl())) {
+      existingConfiguration.setUrl(newMirrorAccessConfiguration.getUrl());
     } else {
       doThrow()
         .violation("url must not be changed", "url")
-        .when(!existingConfig.getUrl().equals(configuration.getUrl()));
-    }
-    if (shouldUpdatePasswordWithExisting(existingConfig, configuration)) {
-      LOG.trace("keeping password for username from existing configuration");
-      configuration.getUsernamePasswordCredential().setPassword(existingConfig.getUsernamePasswordCredential().getPassword());
+        .when(!existingConfiguration.getUrl().equals(newMirrorAccessConfiguration.getUrl()));
     }
-    if (configuration.getCertificateCredential() != null && existingConfig.getCertificateCredential() != null) {
-      if (DUMMY_PASSWORD.equals(configuration.getCertificateCredential().getPassword())) {
-        LOG.trace("keeping password for certificate from existing configuration");
-        configuration.getCertificateCredential().setPassword(existingConfig.getCertificateCredential().getPassword());
+  }
+
+  private void updateCertificateCredentials(MirrorConfiguration existingConfiguration, MirrorAccessConfiguration newMirrorAccessConfiguration) {
+    if (newMirrorAccessConfiguration.getCertificateCredential() != null) {
+      final boolean shouldUpdatePassword = !DUMMY_PASSWORD.equals(newMirrorAccessConfiguration.getCertificateCredential().getPassword());
+      final boolean shouldUpdateCertificate = newMirrorAccessConfiguration.getCertificateCredential().getCertificate() != null;
+      if (existingConfiguration.getCertificateCredential() == null && (shouldUpdatePassword || shouldUpdateCertificate)) {
+        existingConfiguration.setCertificateCredential(new MirrorAccessConfiguration.CertificateCredential());
       }
-      if (configuration.getCertificateCredential().getCertificate() == null) {
-        LOG.trace("keeping certificate from existing configuration");
-        configuration.getCertificateCredential().setCertificate(existingConfig.getCertificateCredential().getCertificate());
+      if (shouldUpdatePassword) {
+        LOG.trace("updating password for certificate");
+        existingConfiguration.getCertificateCredential().setPassword(newMirrorAccessConfiguration.getCertificateCredential().getPassword());
       }
+      if (shouldUpdateCertificate) {
+        LOG.trace("updating certificate");
+        existingConfiguration.getCertificateCredential().setCertificate(newMirrorAccessConfiguration.getCertificateCredential().getCertificate());
+      }
```

Comment:
But it only sets the empty object if the existing configuration does not have a certificateCredential yet. Otherwise, it simply updates the values of the existing certificateCredential. Unless I read it wrong. I'll add a unit test for it to confirm ?

Reply:
You're right, missed the first check.